### ISSUE TITLE
Feature proposal: Run gopass in "native messaging app mode" -> gopass jsonapi

### ISSUE THREAD

**Issue by martinhoefling** [justwatchcom/gopass#306]
This is a proposal for a feature I would like to see (and implement) in gopass. I've shortly talked about the idea with @metalmatze at promcon and he suggested to open an issue for it:

`gopass jsonapi` to communicate via json messages over stdin/stdout.

Calling gopass from the browser requires [native messaging](https://developer.mozilla.org/en-US/Add-ons/WebExtensions/Native_messaging). This would enable usage of gopass from web extensions (browser-plugins), i.e. to fill passwords. 

I've written a prototype for the browser-plugin part similar to [browserpass](https://github.com/dannyvankooten/browserpass). Interfacing with gopass would require the exchange of (json-)messages via stdin/stdout. For this, I see two main solutions:

1. writing a separate program to call gopass with appropriate parameters or "vendor" the gopass code
2. integrating the functionality into gopass itself

While advantages of solution 1. are a decoupled codebase, every change in the gopass (api) requires adaption of the separate program. My preferred way therefore would be to integrate the json messaging api into gopass itself. This would simplify maintainability and avoid breaking changes. While the primary use case I have in mind is running gopass as native messaging app from the browser, this would also ease integration of gopass in other tools.

Two messages `{ "type": "find", "query": "facebook" }` and `{ "type": "getSecret", "label": "gopass/socialmedia/facebook.com" }` as well as their corresponding json responses would be a useful starting point.

What do you think of the idea?

**Comment by dominikschulz** [comment]
TL;DR: I love it. Some concerns implementation wise, but no show stoppers. Go ahead :smile: 

This sounds like a very good and resonable idea that should open up the way for gopass based browser plugins. Amazing. Your reasoning seems sound and I'd also prefer to include this functionality in gopass rather than building a wrapper or similar.

Reading the native messaging specification on MDN it occured to me that calling `gopass jsonapi` may not be possible: 

> The native app is passed two arguments when it starts:
>
> the complete path to the app manifest
> (new in Firefox 55) the ID (as given in the applications manifest.json key) of the add-on that started it.

If I understood this correctly we may either need a simple wrapper that runs `gopass jsonapi` or detect that gopass is called in this way (which would be a little error prone).

The `native messaging` protocol looks a little over engineered but there seems to be straight forward go implementations [available](https://github.com/fuzzdota/chromeapp-native-messaging-sample/blob/master/native/src/cnative/chrome_native.go).

Please let us know if you start working on the gopass part of this.

**Comment by martinhoefling** [comment]
Yes, I'll pick it up. A shell wrapper might be a good idea anyhow: So far I have no other solution to make gpg-agent available to (go)pass started from the browser, since the browser itself does not have the agent settings in its env. This allows me to use my smartcard / yubikey to decrypt the secrets, when the browser tries to access secrets in a certain mount. Further, such a wrapper just does the setup part, no conversion of stdin/stdout I/O, therefore the above reasons for integrating into gopass itself should still apply. I'll work [here](https://github.com/martinhoefling/gopass/tree/jsonapi) on this.

**Comment by dominikschulz** [comment]
Very much looking forward to that!

When using gopass as an backend for a browser extension it may make sense to also think about a (default) storage layout for secrets. Maybe you can/want to have a look at #282

**Comment by martinhoefling** [comment]
#282 is probably a good idea for new users. I personally have a different layout, since most of my secrets are of type "website". My layout is one repo / mount per project. Separate mounts for shared repos with multiple recipients and one repo also synced to my cell phone. The websites are grouped depending on the context. In the project repos, I use "ops" and "development" while in my private repos, secrets are grouped by "SocialMedia", "EMail", "Domains" ... so "website, account, shop, pin" wouldn't be useful for my case.
How would you see such a storage layout affect the json api?

**Comment by dominikschulz** [comment]
The JSON API should not be affected by the structure of the vault, but I was thinking about how to use the JSON API in the browser. If you want good, automatic password lookup (don't know if you were planning for that), the store will probably need to follow a certain structure.

**Comment by martinhoefling** [comment]
I've published the initial code for the browser plugin [here](https://github.com/martinhoefling/gopassbridge). Currently, there is no magic to install the proper app-manifests.

**Comment by martinhoefling** [comment]
@dominikschulz @metalmatze - what are your thoughts about using the gopass icon / logo for the [gopassbridge](https://github.com/martinhoefling/gopassbridge) extension? Shall I create my own or are you fine with using it?

**Comment by dominikschulz** [comment]
I think using the logo should be OK. @falschparker82 WDYT?
